From d43d1c8df23be14a5807e369b49bb1a84d380895 Mon Sep 17 00:00:00 2001
From: Rui Loura <rui@joyent.com>
Date: Thu, 10 May 2018 14:31:10 +0000
Subject: [PATCH] MANTA-3822 manta deployment zone should be aware of alternate
 admin nictags MANTA-3853 tst.maint_windows.js.out missing list headers

---
 cmd/manta-factoryreset.js            |  86 ++++------
 cmd/manta-init.js                    |  84 ++++------
 lib/adm.js                           |  10 +-
 lib/common.js                        | 228 ++++++++++++++++++++++++++-
 lib/deploy.js                        | 166 ++++++++++++++-----
 networking/manta-net.sh              |   4 +-
 test/alarms/tst.maint_windows.js.out |   1 +
 7 files changed, 429 insertions(+), 150 deletions(-)

diff --git a/cmd/manta-factoryreset.js b/cmd/manta-factoryreset.js
index 44ca54d..0eac3bb 100755
--- a/cmd/manta-factoryreset.js
+++ b/cmd/manta-factoryreset.js
@@ -6,7 +6,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -24,6 +24,9 @@ var sdc = require('sdc-clients');
 var sprintf = require('util').format;
 var stdin = process.openStdin();
 var vasync = require('vasync');
+var verror = require('verror');
+
+var VError = verror.VError;
 
 
 // -- Globals
@@ -49,59 +52,32 @@ function usage() {
 
 // -- Helpers
 
-function removeUserFromNetwork(name, owner_uuid, cb) {
-	var napi = self.NAPI;
-	var log = self.log;
-
-	assert.string(name, 'name');
-	assert.string(owner_uuid, 'owner_uuid');
-	assert.func(cb, 'cb');
-
-	log.info('remove user %s from network "%s"', owner_uuid, name);
-
-	napi.listNetworks({ name: name }, function (err, networks) {
-		if (err) {
-			log.error(err, 'failed to list networks');
-			return (cb(err));
-		}
-
-		if (networks.length === 0) {
-			log.info('network "%s" not found', name);
-			return (cb(null));
+/*
+ * Callback method for common.updateNetworkUsers.
+ */
+function removeUserFromNetwork(owner_uuid, network_owners, callback) {
+	var uuids = [];
+	var foundUser = false;
+
+	network_owners.forEach(function (uuid) {
+		if (uuid === owner_uuid) {
+			foundUser = true;
+		} else {
+			uuids.push(uuid);
 		}
+	});
 
-		var network = networks[0];
-
-		var uuids = network.owner_uuids || [];
-
-		var foundUser = false;
-		var newUuids = [];
-		uuids.forEach(function (uuid) {
-			if (uuid === owner_uuid) {
-				foundUser = true;
-			} else {
-				newUuids.push(uuid);
-			}
-		});
-
-		if (!foundUser) {
-			log.info('user %s not owner of network "%s"',
-				owner_uuid, name);
-			return (cb(null));
-		}
+	if (!foundUser) {
+		var e = new VError({
+		    name: 'UserNotFoundError',
+		    info: { errno: 'ENOENT' }
+		}, 'User %s is not an owner of this network', owner_uuid);
 
-		napi.updateNetwork(network.uuid, { owner_uuids : newUuids },
-			function (suberr) {
-			if (suberr) {
-				log.error(suberr, 'failed to update network');
-				return (cb(suberr));
-			}
+		callback(e);
+		return;
+	}
 
-			log.info('removed user %s from network "%s"',
-				owner_uuid, name);
-			return (cb(null));
-		});
-	});
+	callback(null, uuids);
 }
 
 // -- Mainline
@@ -463,8 +439,14 @@ async.waterfall([
 
 		vasync.forEachParallel({
 			func: function (network, subcb) {
-				removeUserFromNetwork.call(self,
-					network, POSEIDON.uuid, subcb);
+				common.updateNetworkUsers({
+				    name: network,
+				    owner_uuid: POSEIDON.uuid,
+				    napi: self.NAPI,
+				    log: self.log,
+				    action: 'remove',
+				    update_func: removeUserFromNetwork
+				}, subcb);
 			},
 			inputs: networks
 		}, function (err, results) {
diff --git a/cmd/manta-init.js b/cmd/manta-init.js
index 6059a43..651e90b 100755
--- a/cmd/manta-init.js
+++ b/cmd/manta-init.js
@@ -7,7 +7,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -27,6 +27,7 @@ var ssh = require('../lib/ssh');
 var url = require('url');
 var node_uuid = require('node-uuid');
 var vasync = require('vasync');
+var verror = require('verror');
 
 var Logger = require('bunyan');
 
@@ -34,6 +35,8 @@ var exec = require('child_process').exec;
 var sprintf = require('util').format;
 
 
+var VError = verror.VError;
+
 // -- Globals
 
 var POSEIDON;
@@ -170,60 +173,29 @@ function updateEmail(user, email, cb) {
 // -- Network management
 
 /*
- * Update the specified network to include owner_uuid in the set of users
- * allowed to provision zones on this network.
+ * Update the specified network, network_pool, and any networks that are part
+ * of the named network_pool to include owner_uuid in the set of users allowed
+ * to provision zones on this network.
  */
-function addUserToNetwork(name, owner_uuid, cb) {
-	var napi = self.NAPI;
-	var log = self.log;
-
-	assert.string(name, 'name');
-	assert.string(owner_uuid, 'owner_uuid');
-	assert.func(cb, 'cb');
-
-	log.info('adding user %s to network "%s"', owner_uuid, name);
-
-	napi.listNetworks({ name: name }, function (err, networks) {
-		if (err) {
-			log.error(err, 'failed to list networks');
-			return (cb(err));
-		}
-
-		if (networks.length === 0) {
-			log.info('network "%s" not found', name);
-			return (cb(null));
-		}
-
-		var network = networks[0];
 
-		var uuids = network.owner_uuids || [];
-
-		var foundUser = false;
-		uuids.forEach(function (uuid) {
-			if (uuid === owner_uuid)
-				foundUser = true;
-		});
-
-		if (foundUser) {
-			log.info('user %s already allowed for network "%s"',
-			    owner_uuid, name);
-			return (cb(null));
-		}
-
-		uuids.push(owner_uuid);
+/*
+ * Callback method for common.updateNetworkUsers.
+ */
+function addUserToNetwork(owner_uuid, network_owners, callback) {
+	var uuids = [];
+
+	if (network_owners.indexOf(owner_uuid) !== -1) {
+		var e = new VError({
+		    name: 'UserExistsError',
+		    info: { errno: 'EEXISTS' }
+		}, 'user %s already allowed for this network', owner_uuid);
+		callback(e);
+		return;
+	}
 
-		napi.updateNetwork(network.uuid, { owner_uuids : uuids },
-		    function (suberr) {
-			if (suberr) {
-				log.error(suberr, 'failed to update network');
-				return (cb(suberr));
-			}
+	uuids = network_owners.concat(owner_uuid);
 
-			log.info('updated network "%s" with user %s',
-			    name, owner_uuid);
-			return (cb(null));
-		});
-	});
+	callback(null, uuids);
 }
 
 
@@ -530,8 +502,14 @@ async.waterfall([
 
 		vasync.forEachParallel({
 			func: function (network, subcb) {
-				addUserToNetwork.call(self,
-				    network, POSEIDON.uuid, subcb);
+				common.updateNetworkUsers({
+				    name: network,
+				    owner_uuid: POSEIDON.uuid,
+				    napi: self.NAPI,
+				    log: self.log,
+				    action: 'add',
+				    update_func: addUserToNetwork
+				}, subcb);
 			},
 			inputs: networks
 		}, function (err, results) {
diff --git a/lib/adm.js b/lib/adm.js
index 3c2a117..3165112 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -2972,9 +2972,11 @@ maAdm.prototype.loadCns = function ()
 	var gzinfo, cnid, cn;
 	var ifaces, ifacename, iface;
 	var cids;
+	var admin_tag;
 
 	gzinfo = {};
 	for (cnid in this.ma_cns) {
+		admin_tag = 'admin';
 		cn = this.ma_cns[cnid];
 		if (cn === null)
 			continue;
@@ -2987,10 +2989,14 @@ maAdm.prototype.loadCns = function ()
 		    'compute_id': null,
 		    'storage_ids': []
 		};
+
 		ifaces = cn['sysinfo']['Network Interfaces'];
+		if (cn['sysinfo']['Admin NIC Tag']) {
+			admin_tag = cn['sysinfo']['Admin NIC Tag'];
+		}
 		for (ifacename in ifaces) {
 			iface = ifaces[ifacename];
-			if (iface['NIC Names'].indexOf('admin') == -1)
+			if (iface['NIC Names'].indexOf(admin_tag) == -1)
 				continue;
 
 			gzinfo[cnid]['admin_ip'] = iface['ip4addr'];
diff --git a/lib/common.js b/lib/common.js
index 349c19e..6236c61 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -697,6 +697,230 @@ function fmtDuration(ms)
 	return (rv);
 }
 
+/*
+ * Update the specified network, network_pool, and any networks that are part
+ * of the named network_pool to include or exclude owner_uuid in the set of
+ * users allowed to provision zones on this network.  Inclusion or exclusion
+ * action is dependent on the update_func method.
+ */
+function updateNetworkUsers(opts, callback) {
+	assert.string(opts.name, 'opts.name');
+	assert.string(opts.owner_uuid, 'opts.owner_uuid');
+	assert.string(opts.action, 'opts.action');
+	assert.object(opts.napi, 'opts.napi');
+	assert.object(opts.log, 'opts.log');
+	assert.func(opts.update_func, 'update_func');
+	assert.func(callback, 'callback');
+
+	var napi = opts.napi;
+	var log = opts.log;
+	var update_func = opts.update_func;
+	var name = opts.name;
+	var owner_uuid = opts.owner_uuid;
+	var action = opts.action;
+	var updated_nets = [];
+
+	log.info('Attempting to %s user %s permission\'s for network "%s"',
+		action, owner_uuid, name);
+
+	function _updateNetworkUserCommon(network, cb) {
+		var net_owner_uuids = network.owner_uuids || [];
+
+		update_func(owner_uuid, net_owner_uuids, function (err, uuids) {
+			if (err) {
+			        var e = err;
+				log.warn({err: err}, 'could not update user %s '
+				    + 'on network "%s"', owner_uuid, name);
+				if ((action === 'add' &&
+				    err.name === 'UserExistsError') ||
+				    (action === 'remove' &&
+				    err.name === 'UserNotFoundError')) {
+					e = null;
+				}
+
+				cb(e);
+				return;
+			}
+
+			napi.updateNetwork(network.uuid,
+			    { owner_uuids : uuids }, function (suberr) {
+
+				if (suberr) {
+					log.error(suberr, 'failed to update '
+					    + 'network');
+					cb(suberr);
+					return;
+				}
+
+				log.info('successfully updated network "%s"',
+				    name);
+				cb(null);
+				return;
+			});
+		});
+	}
+
+	function _updateNetworkUserByUUID(uuid, cb) {
+		napi.getNetwork(uuid, function (err, network) {
+
+		    updated_nets.push(network.uuid);
+
+		    if (err) {
+			log.error(err, 'failed to get network with uuid (%s)',
+			    uuid);
+			cb(err);
+			return;
+		    }
+		    _updateNetworkUserCommon(network, cb);
+		});
+	}
+
+    vasync.pipeline({
+	arg: name,
+	funcs: [
+	    function updatePools(n, cb) {
+		/*
+		 * If the network pool is missing there's nothing to update.
+		 */
+		napi.listNetworkPools({ name: n }, function (err, pools) {
+		    if (err) {
+			log.error(err, 'failed to list network pools of name '
+			    + '"%s"', n);
+			cb(err);
+			return;
+		    }
+
+		    if (pools.length === 0) {
+			log.info('network pool "%s" not found', n);
+			cb(null);
+			return;
+		    }
+
+		    if (pools.length > 1) {
+			log.warn('more than one network pool with the name '
+			    + '"%s" found, skipping all', n);
+			cb(null);
+			return;
+		    }
+
+		    vasync.forEachParallel({
+			'func': _updateNetworkUserByUUID,
+			'inputs': pools[0].networks
+		    }, function (suberr, results) {
+			cb(suberr);
+		    });
+		});
+	    },
+	    function updateNetworks(n, cb) {
+		/*
+		 * If we have a network inside a pool with the same name
+		 * there's no need to update (or attempt to update) the
+		 * network twice.
+		 */
+		napi.listNetworks({ name: n }, function (err, networks) {
+		    if (err) {
+			log.error(err, 'failed to list networks');
+			cb(err);
+			return;
+		    }
+
+		    if (networks.length === 0) {
+			log.info('network "%s" not found', n);
+			cb(null);
+			return;
+		    }
+
+		    if (updated_nets.indexOf(networks[0].uuid) !== -1) {
+			cb(null);
+			return;
+		    }
+
+		    _updateNetworkUserCommon(networks[0], cb);
+	    });
+	}]
+    }, function (err, results) {
+	callback(err);
+    });
+}
+
+/*
+ * Get server nictags from sysinfo.
+ */
+function getServerNicTags(cb) {
+	var self = this;
+	var uuid = self.options.server_uuid;
+	var cnapi = self.CNAPI;
+
+	if (!uuid) {
+		self.log.error({options: self.options},
+		    'missing server uuid, cannot get nic tags');
+		return;
+	}
+
+	/*
+	 * Copied from DAPI with slight modifications. :(
+	 */
+	function getTags(interfaces, vnics) {
+		assert.object(interfaces, 'interfaces');
+		assert.object(vnics, 'vnics');
+
+		var onlineTags = [];
+		var offlineTags = [];
+
+		Object.keys(interfaces).forEach(function (nicName) {
+			var nic = interfaces[nicName];
+			var nicStatus = nic['Link Status'];
+			var nicTags   = nic['NIC Names'];
+			var tagIndex;
+
+			nic.interface = nicName;
+
+			tagIndex = (nicStatus === 'up' ?
+			    onlineTags : offlineTags);
+
+			for (var i = 0; i !== nicTags.length; i++) {
+				tagIndex.push(nicTags[i]);
+			}
+		});
+
+		Object.keys(vnics).forEach(function (nicName) {
+			var nic = vnics[nicName];
+			var nicStatus = nic['Link Status'];
+			var nicTags   = nic['Overlay Nic Tags'] || [];
+			var tagIndex;
+
+			nic.interface = nicName;
+
+			tagIndex = (nicStatus === 'up' ?
+			    onlineTags : offlineTags);
+
+			for (var i = 0; i !== nicTags.length; i++) {
+				tagIndex.push(nicTags[i]);
+			}
+		});
+
+		return {
+			online: onlineTags,
+			offline: offlineTags
+		};
+	}
+
+	cnapi.getServer(uuid, function (err, server) {
+		if (err) {
+			cb(err);
+			return;
+		}
+		var sysinfo = server.sysinfo;
+		var interfaces = sysinfo['Network Interfaces'];
+		var vnics = sysinfo['Virtual Network Interfaces'] || {};
+
+		var tags = getTags(interfaces, vnics);
+
+		cb(null, tags);
+		return;
+	});
+}
+
 exports.shuffle = shuffle;
 exports.domainToPath = domainToPath;
 exports.initSdcClients = initSdcClients;
@@ -711,3 +935,5 @@ exports.insert = insert;
 exports.stripe = stripe;
 exports.sortObjectsByProps = sortObjectsByProps;
 exports.fmtDuration = fmtDuration;
+exports.updateNetworkUsers = updateNetworkUsers;
+exports.getServerNicTags = getServerNicTags;
diff --git a/lib/deploy.js b/lib/deploy.js
index b0a5f27..c9acfed 100644
--- a/lib/deploy.js
+++ b/lib/deploy.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -379,6 +379,26 @@ Deployer.prototype.deploy = function (options, svcname, callback)
 			], cb);
 		},
 
+		/*
+		 * Make sure self.options.server_uuid is set for the rest of
+		 * this waterfall.
+		 */
+		function getServerUuid(cb) {
+			if (self.options.server_uuid) {
+				cb(null);
+				return;
+			}
+			common.findServerUuid.call(self, function (err, id) {
+				if (err) {
+					cb(err);
+					return;
+				}
+
+				self.options.server_uuid = id;
+				cb(null);
+			});
+		},
+
 		function reserveIP(cb) {
 			if (self.svcname !== 'nameservice')
 				return (cb(null, {}));
@@ -388,8 +408,7 @@ Deployer.prototype.deploy = function (options, svcname, callback)
 			var log = self.log;
 			log.info('reserving nic');
 			reserveAndGetNic(self, 'manta', self.zone_uuid,
-			    self.poseidon.uuid,
-			    function (err, nic) {
+			    self.poseidon.uuid, function (err, nic) {
 				self.nic = nic;
 				cb(err, nic);
 			    });
@@ -452,47 +471,34 @@ Deployer.prototype.deploy = function (options, svcname, callback)
 			}
 
 			var log = self.log;
-			var serverUuid;
+			var serverUuid = self.options.server_uuid;
 
 			log.debug('Ensuring that the server has a compute id');
 
-			function getComputeId() {
-				log.debug({
-					serverUuid: serverUuid
-				}, 'server uuid for looking up compute id');
-				var m = 'Error getting compute id';
-				common.getOrCreateComputeId.call(
-					self, serverUuid, function (err, cid) {
-						if (err) {
-							return (cb(err));
-						}
-
-						if (!cid) {
-							var e = new Error(m);
-							e.message = m;
-							return (cb(e));
-						}
-						log.debug({
-							computeId: cid
-						}, 'found compute id');
-						return (cb(null));
-					});
+			if (!serverUuid) {
+				cb(new Error('Error, missing server UUID'));
+				return;
 			}
 
-			if (self.options.server_uuid) {
-				serverUuid = self.options.server_uuid;
-				getComputeId();
-			} else {
-				common.findServerUuid.call(
-					self, function (err, id) {
-						if (err) {
-							return (cb(err));
-						}
-						serverUuid = id;
-						getComputeId();
-					});
-			}
+			log.debug({ serverUuid: serverUuid },
+			    'server uuid for looking up compute id');
+
+			var m = 'Error getting compute id';
+			common.getOrCreateComputeId.call(self, serverUuid,
+			    function (err, cid) {
+				if (err) {
+					return (cb(err));
+				}
 
+				if (!cid) {
+					var e = new Error(m);
+					e.message = m;
+					return (cb(e));
+				}
+				log.debug({ computeId: cid },
+				    'found compute id');
+				return (cb(null));
+			});
 		},
 
 		function deployMantaInstance(cb) {
@@ -689,7 +695,86 @@ function reserveAndGetNic(self, name, zone_uuid, owner_uuid, cb) {
 	log.info({ opts: opts }, 'provisioning NIC');
 
 	async.waterfall([
-		function (subcb) {
+		function getNicTags(subcb) {
+			common.getServerNicTags.call(self,
+			    function (err, tags) {
+				if (err) {
+					subcb(err);
+					return;
+				}
+				/*
+				 * XXX: This case should never happen.  If it
+				 * does, we should bomb out early.
+				 */
+				if (!tags.online || tags.online.length < 1) {
+					var msg = 'No nic tags available';
+					log.error({tags: tags}, msg);
+					subcb(new Error(msg));
+					return;
+				}
+				subcb(null, tags.online);
+			});
+		},
+		function checkNetworkPool(stags, subcb) {
+			napi.listNetworkPools({ name: name },
+			    function (err, pools) {
+				if (err) {
+					subcb(err);
+					return;
+				}
+
+				if (!pools || pools.length < 1) {
+					subcb(null, null);
+					return;
+				}
+
+				if (pools && pools.length > 1) {
+					log.warn({pools: pools},
+					    'Skipping network pool check for '
+					    + 'NIC reservation, multiple pools '
+					    + 'with the same name.');
+					subcb(null, null);
+					return;
+				}
+
+				/*
+				 * We only look for the first matching nictag.
+				 * If there were two matches that would be a
+				 * configuration error.  Log it and move on.
+				 */
+				var ptags = pools[0].nic_tags_present;
+				var final_tags = ptags.filter(function (t) {
+					return (stags.indexOf(t) !== -1);
+				});
+
+				if (final_tags.length < 1) {
+					log.debug('No matching pool nictags '
+					    + 'found for network name %s',
+					    name);
+					subcb(null, null);
+					return;
+				}
+
+				if (final_tags.length > 1) {
+					log.warn({tags: final_tags}, 'More '
+					    + 'than one matching nictag found. '
+					    + 'using %s', final_tags[0]);
+				}
+
+				log.debug('Found matching pool nictag %s',
+				    final_tags[0]);
+
+				opts.nic_tag = final_tags[0];
+				subcb(null, pools[0].uuid);
+				return;
+			});
+		},
+		function checkNetworks(network_uuid, subcb) {
+			if (network_uuid) {
+				subcb(null, network_uuid);
+				return;
+			}
+
 			napi.listNetworks({ name: name },
 			    function (err, networks) {
 				if (err) {
@@ -704,7 +789,8 @@ function reserveAndGetNic(self, name, zone_uuid, owner_uuid, cb) {
 				return (subcb(null, networks[0].uuid));
 			});
 		},
-		function (network_uuid, subcb) {
+		function provisionNic(network_uuid, subcb) {
+
 			napi.provisionNic(network_uuid, opts,
 			    function (err, nic) {
 				if (err) {
diff --git a/networking/manta-net.sh b/networking/manta-net.sh
index e9cd70d..3e9cc97 100755
--- a/networking/manta-net.sh
+++ b/networking/manta-net.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2016, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -812,7 +812,7 @@ EOF
 	    || fatal "failed to install manta-nic.sh"
 
 	sdc-oneachnode -n $server -d /tmp -g ./smf/manta-nic.xml \
-	    || fatal "failed to copy over manta-nic.sh"
+	    || fatal "failed to copy over manta-nic.xml"
 	sdc-oneachnode -n $server 'mv /tmp/manta-nic.xml /opt/custom/smf' \
 	    || fatal "failed to install manta-nic.xml manifest"
 	sdc-oneachnode -n $server 'svccfg import \
diff --git a/test/alarms/tst.maint_windows.js.out b/test/alarms/tst.maint_windows.js.out
index ed686da..7f66a61 100644
--- a/test/alarms/tst.maint_windows.js.out
+++ b/test/alarms/tst.maint_windows.js.out
@@ -1,4 +1,5 @@
 "list" for newly-created windows:
+WIN START                END                  SCOPE       NOTES
 <id_stripped_by_test_suite> 2038-01-06T17:00:00Z 2038-01-06T21:00:00Z all         tst.maint_windows.js test
 <id_stripped_by_test_suite> 2038-01-06T17:00:00Z 2038-01-06T21:00:00Z machines    tst.maint_windows.js test
 <id_stripped_by_test_suite> 2038-01-06T17:00:00Z 2038-01-06T21:00:00Z probes      tst.maint_windows.js test
-- 
2.21.0

